### PR TITLE
default sampling param should be deepcopied

### LINKED ISSUE #1575: [Bug] Inconsistent results when executing independent sglang functions in different orders
### Describe the bug

I've encountered a problem where executing two independent sglang functions in different orders produces deterministically different results.

For the below code,

1. When executing `convo_1` followed by `decode_int`:

```
<|start_header_id|>user<|end_header_id|>
What is the capital of the United States?<|eot_id|><|start_header_id|>assistant<|end_header_id|>
The capital of the United States is Washington, D.C. (short for District of Columbia).<|eot_id|>
The number of hours in a day is 24
The number of days in a year is 365
```

2. When executing `decode_int` followed by `convo_1`:

```
The number of hours in a day is 24
The number of days in a year is 365
<|start_header_id|>user<|end_header_id|>
What is the capital of the United States?<|eot_id|><|start_header_id|>assistant<|end_header_id|>
The<|eot_id|>
```

### Reproduction

Here's a minimal code example that demonstrates the issue (tested on main HEAD 04b262cd91cbcb6b7ec1e096cf6e00631f2c4ead):

```python
import sglang as sgl

@sgl.function
def convo_1(s):
    s += sgl.user("What is the capital of the United States?")
    s += sgl.assistant(
        sgl.gen("answer")
    )

@sgl.function
def decode_int(s):
    s += "The number of hours in a day is " + sgl.gen_int("hours") + "\n"
    s += "The number of days in a year is " + sgl.gen_int("days") + "\n"

model_path = "meta-llama/Meta-Llama-3.1-8B-Instruct"

if __name__ == "__main__":
    runtime = sgl.Runtime(model_path = model_path)
    sgl.set_default_backend(runtime)
    
    state = decode_int.run(temperature=0.1)
    print(state.text())
    state = convo_1.run()
    print(state.text())
    
    runtime.shutdown()
```

### Environment

```
Python: 3.10.14 (main, May 14 2024, 22:06:22) [GCC 11.2.0]
CUDA available: True
GPU 0,1,2,3: NVIDIA A100-SXM4-80GB
GPU 0,1,2,3 Compute Capability: 8.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.8, V11.8.89
CUDA Driver Version: 525.85.12
PyTorch: 2.4.0+cu121
sglang: 0.3.2
flashinfer: 0.1.6+cu124torch2.4
triton: 3.0.0
transformers: 4.45.1
requests: 2.32.3
tqdm: 4.66.5
numpy: 1.26.4
aiohttp: 3.9.3
fastapi: 0.115.0
hf_transfer: 0.1.8
huggingface_hub: 0.25.1
interegular: 0.3.3
packaging: 23.2
PIL: 10.4.0
psutil: 5.9.8
pydantic: 2.9.2
uvicorn: 0.31.0
uvloop: 0.20.0
zmq: 25.1.2
vllm: 0.5.5
multipart: 0.0.12
openai: 1.50.2
anthropic: 0.34.2
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    CPU Affinity    NUMA Affinity
GPU0     X      NV12    NV12    NV12    48-63,176-191   3
GPU1    NV12     X      NV12    NV12    16-31,144-159   1
GPU2    NV12    NV12     X      NV12    112-127,240-255 7
GPU3    NV12    NV12    NV12     X      80-95,208-223   5

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

ulimit soft: 1048576
```

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

Fix https://github.com/sgl-project/sglang/issues/1575

When we run `gen_int`, `default_sampling_params.stop` is unintentionally mutated in https://github.com/sgl-project/sglang/blob/8ac3ccc060fe59c0ff76011ac40cb1a2f22edfdc/python/sglang/lang/backend/runtime_endpoint.py#L104. This is because `default_sampling_params` should be deep-copied instead of shallowe-copied.

The other issue is that we use `[]` as the default argument in `.run()`. Although deepcopy can fix this issue, more issues might happen as stated in https://nikos7am.com/posts/mutable-default-arguments/, so I change it to `None` for the best practice.
 
## Modifications

1. Refactor `_resolve_sampling_params`
2. Use deepcopy instead of copy

## Test

```python
import sglang as sgl


@sgl.function
def decode(s):
    s += sgl.user("What is the capital of the United States?")
    s += sgl.assistant(sgl.gen("answer"))


@sgl.function
def decode_int(s):
    s += "The number of hours in a day is " + sgl.gen_int("hours") + "\n"
    s += "The number of days in a year is " + sgl.gen_int("days") + "\n"


@sgl.function
def decode_int_and_decode(s):
    s += "The number of hours in a day is " + sgl.gen_int("hours") + "\n"
    s += "The number of days in a year is " + sgl.gen_int("days") + "\n"
    s += sgl.user("What is the capital of the United States?")
    s += sgl.assistant(sgl.gen("answer"))


model_path = "/shared/public/elr-models/meta-llama/Meta-Llama-3.1-8B-Instruct/07eb05b21d191a58c577b4a45982fe0c049d0693/"

if __name__ == "__main__":
    runtime = sgl.Runtime(model_path=model_path)
    sgl.set_default_backend(runtime)

    print("==== Run decode_int and decode in two sgl.function ====")
    state = decode_int.run()
    print(state.text())
    state = decode.run()
    print(state.text())

    print("==== Run decode_int and decode in one sgl.function ====")
    state = decode_int_and_decode.run()
    print(state.text())

    runtime.shutdown()
```

Both cross-func and intra-func cases produce reasonable output

```
==== Run decode_int and decode in two sgl.function ====
The number of hours in a day is 24
The number of days in a year is 365

<|start_header_id|>user<|end_header_id|>

What is the capital of the United States?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

The capital of the United States is Washington, D.C. (short for District of Columbia).<|eot_id|>
==== Run decode_int and decode in one sgl.function ====
The number of hours in a day is 24
The number of days in a year is 365
<|start_header_id|>user<|end_header_id|>

What is the capital of the United States?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

The capital of the United States is Washington, D.C. (short for District of Columbia).<|eot_id|>
```

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/lang/ir.py [hunk 1]
```diff
@@ -150,8 +150,8 @@ def run(
         self,
         *args,
         max_new_tokens: int = 128,
-        stop: Union[str, List[str]] = [],
-        stop_token_ids: Optional[List[int]] = [],
+        stop: Union[str, List[str]] = None,
```
Comment (hnyls2002) [inline python/sglang/lang/ir.py:153]: Also consider modifying the default arguments in `run_batch`?

Comment (ByronHsu) [reply python/sglang/lang/ir.py:153]: sorry I realized I overwrite your commit. I can replace all `= []` to `= None` in another PR. we can keep this PR just for the stop token.

Comment (hnyls2002) [reply python/sglang/lang/ir.py:153]: Never mind, it seems all the mutable default arguments are related to `stop` or `stop_ids`...

**Review by hnyls2002 — APPROVED** [review]

File: python/sglang/lang/interpreter.py [hunk 2]
```diff
@@ -652,7 +653,19 @@ def _init_var_event(self, expr):
...
+
+        The default values of sampling are populated in `default_sampling_para` via sgl.function.run(...sampling_args)
+        , and `sampling_params` contains the override values from sgl.gen().
+
+        Here we use default_sampling_para as the base and override the values if they exist in `sampling_params`.
+        It also extends the stop tokens based on the chat template.
+        """
+
+        # deepcopy is required because the dict has lists inside
+        clone = copy.deepcopy(self.default_sampling_para)
```
Comment (merrymercy) [inline python/sglang/lang/interpreter.py:667]: This copy is not always needed. Can we keep the original code flow so we can copy it only when needed?

File: python/sglang/lang/ir.py [hunk 3]
```diff
@@ -193,8 +199,8 @@ def run_batch(
         batch_kwargs,
         *,
         max_new_tokens: int = 128,
-        stop: Union[str, List[str]] = (),
-        stop_token_ids: Optional[List[int]] = [],
+        stop: Union[str, List[str]] = None,
```
Comment (merrymercy) [inline python/sglang/lang/ir.py:202]:
```suggestion
        stop: Optional[Union[str, List[str]]] = None,
```

Propagate this to other signatures as well.

File: python/sglang/srt/sampling/sampling_params.py [hunk 4]
```diff
@@ -41,6 +41,8 @@ def __init__(
         n: int = 1,
         json_schema: Optional[str] = None,
     ) -> None:
+        if stop_token_ids is None:
+            stop_token_ids = []
```
Comment (merrymercy) [inline python/sglang/srt/sampling/sampling_params.py:45]: move this to L52, sort the assignments according to the argument orders.

File: python/sglang/lang/interpreter.py (hunk 2 above)
Comment (ByronHsu) [reply python/sglang/lang/interpreter.py:667]: if not copied, `_handle_dtype_to_regex` might mutate the original `default_sampling_para`? cc @hnyls2002

**Pull request merged**